backend/controllers/auth-controller.js

```diff
@@ -0,0 +1,75 @@
+import { Prisma } from "../config/prisma.js"
+import bcrypt from 'bcrypt'
+import jwt from 'jsonwebtoken'
+
+export const register = async (req, res) => {
+    try {
+        const { username, email, password } = req.body
+
+        const existUsername = await Prisma.user.findFirst({ where: { username } })
+        if (existUsername) {
+            return res.status(400).json({ message: 'Username already exist' })
+        }
+
+        const existEmail = await Prisma.user.findFirst({ where: { email } })
+        if (existEmail) {
+            return res.status(400).json({ message: 'Email already exist' })
+        }
+
+        const hashPassword = await bcrypt.hash(String(password), parseInt(process.env.SALT_ROUND))
+
+        const registerUser = await Prisma.user.create({
+            data: {
+                username,
+                email,
+                password: hashPassword
+            }
+        })
+
+        if (!registerUser) {
+            return res.status(500).json({ message: 'User register fail' })
+        }
+
+        return res.json({ message: 'Register success' })
+
+    } catch (err) {
+        return res.status(500).json({ message: 'Register error', error: err.message })
+    }
+}
+
+export const login = async (req, res) => {
+    try {
+        const { email, password } = req.body
+
+        const existUser = await Prisma.user.findFirst({ where: { email } })
+
+        if (!existUser) {
+            return res.status(400).json({ message: 'Incorrect email or password' })
+        }
+
+        const passwordMatch = await bcrypt.compare(String(password), existUser.password)
+
+        if (!passwordMatch) {
+            return res.status(400).json({ message: 'Incorrect email or password' })
+        }
+
+        let payload = {
+            id: existUser.id,
+            username: existUser.username,
+            role: existUser.role
```

Comment:
Please note that there is an alternative way to implement this role.

It's returning id or username without role. Then when request come to backend, we get role by using middleware and attach to `req.user`. It's more secure. However, it trades with latency on the backend.

Reply:
Yes, that's what I meant. One system can have different implementation. This is just one possible way.